packages/lab/src/list-next/useList.ts

```diff
@@ -0,0 +1,284 @@
+    return optionIndex !== -1 ? optionIndex : -1;
+  };
+
+  const focusAndSelect = (element: Element) => {
+    setSelectedIndexes([getListItemIndex(element)]);
+    setActiveDescendant(element.id);
+    setFocusedIndex(getListItemIndex(element));
+    updateScroll(element);
+
+    if (onFocus) {
+      onFocus(element);
+    }
+  };
+
+  const focusFirstItem = () => {
+    // Find first active item
+    const firstItem = activeOptions[0];
+    if (firstItem) {
+      focusAndSelect(firstItem);
+    }
+  };
+
+  const focusLastItem = () => {
+    // Find last active item
+    const lastItem = activeOptions[activeOptions.length - 1];
+    if (lastItem) {
+      focusAndSelect(lastItem);
+    }
+  };
+
+  const toggleSelectItem = (element: Element) => {
+    const elementIndex = getListItemIndex(element);
+    const itemIsSelected = selectedIndexes.includes(elementIndex);
+    const newSelection = deselectable && itemIsSelected ? [] : [elementIndex];
+    setSelectedIndexes(newSelection);
+  };
+
+  const justFocusItem = (element: Element) => {
+    if (onFocus) {
+      onFocus(element);
+    }
+    setActiveDescendant(element.id);
+    setFocusedIndex(getListItemIndex(element));
+    updateScroll(element);
+  };
+
+  const findNextOption = (
+    currentOption: Element | null,
+    moves: number
+  ): Element => {
+    // Returns next item, if no current option it will return 0
+    const nextOptionIndex = currentOption
+      ? activeOptions.indexOf(currentOption) + moves
+      : 0;
+    return (
+      activeOptions[nextOptionIndex] || activeOptions[activeOptions.length - 1]
+    );
+  };
+
+  const findPreviousOption = (
+    currentOption: Element,
+    moves: number
+  ): Element => {
+    // Return the previous option if it exists; otherwise, returns first option
+    const currentOptionIndex = activeOptions.indexOf(currentOption);
+    return activeOptions[currentOptionIndex - moves] || activeOptions[0];
+  };
+
+  const updateScroll = (currentTarget: Element) => {
+    if (!list || !currentTarget) return;
+    const { offsetTop, offsetHeight } = currentTarget;
+    const listHeight = list?.clientHeight;
+    const listScrollTop = list?.scrollTop;
+    if (offsetTop < listScrollTop) {
+      list.scrollTop = offsetTop;
+    } else if (offsetTop + offsetHeight > listScrollTop + listHeight) {
+      list.scrollTop = offsetTop + offsetHeight - listHeight;
+    }
+  };
+
+  // Handlers
+  const handleClick = ({ currentTarget }: MouseEvent<HTMLUListElement>) => {
+    const nonClickableTarget = activeOptions.indexOf(currentTarget) === -1;
+    if (nonClickableTarget) {
+      return;
+    }
+    toggleSelectItem(currentTarget);
+    setActiveDescendant(currentTarget.id);
+    setFocusedIndex(null);
+    updateScroll(currentTarget);
+  };
+
+  const handleBlur = useEventCallback(() => {
+    setFocusedIndex(null);
+    if (onBlur) {
+      onBlur();
+    }
+  });
+
+  const handleMouseDown = useEventCallback(() => {
```

Comment:
this handler is needed to know if on focus the event comes from a click or the keyboard, if it doesn't come from the keyboard it doesn't add the visual focus ring.